platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractCollectionProperty.java

```diff
@@ -205,8 +206,20 @@ public int size() {
      */
     private void addExplicitCollector(Collector<T> collector) {
         assertCanMutate();
-        CollectionSupplier<T, C> explicitValue = getExplicitValue(defaultValue);
-        setSupplier(explicitValue.plus(collector));
+        setSupplier(withAppendedValue(collector));
+    }
+
+    private CollectionSupplier<T, C> withAppendedValue(Collector<T> value) {
+        CollectionSupplier<T, C> currentValue = getExplicitValue(defaultValue);
+        try {
+            return currentValue.plus(value);
+        } catch (IllegalStateException e) {
+            throw new ConcurrentModificationException(
+                "State of " + getDisplayName().getDisplayName() + " is corrupted. " +
+                    "This may be caused by unsafe concurrent modifications with parallel configuration or execution enabled.",
```

Comment:
💅  Would it make sense to have this message defined in a single place, so it's always consistently updated, if it ever needs to change?

Reply:
I've moved it to AbstractProperty